src/Microsoft.Azure.WebJobs.Host/Loggers/Logger/Aggregator/FunctionResultAggregator.cs

```diff
@@ -0,0 +1,125 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.WebJobs.Host.Loggers
+{
+    internal class FunctionResultAggregator : IFunctionResultAggregator, IDisposable
+    {
+        private readonly ILogger _logger;
+        private readonly int _batchSize;
+        private readonly TimeSpan _batchTimeout;
+
+        private BufferBlock<FunctionResultLog> _buffer;
+        private BatchBlock<FunctionResultLog> _batcher;
+
+        private Timer _windowTimer;
+        private IDisposable[] _disposables;
+
+        public FunctionResultAggregator(int batchSize, TimeSpan batchTimeout, ILoggerFactory loggerFactory)
+        {
+            _logger = loggerFactory.CreateLogger(LoggingCategories.Aggregator);
+            _batchSize = batchSize;
+            _batchTimeout = batchTimeout;
+            InitializeFlow(_batchSize, _batchTimeout);
+        }
+
+        private void InitializeFlow(int maxBacklog, TimeSpan maxFlushInterval)
+        {
+            _buffer = new BufferBlock<FunctionResultLog>(
+                new ExecutionDataflowBlockOptions()
+                {
+                    BoundedCapacity = maxBacklog
+                });
+
+            _batcher = new BatchBlock<FunctionResultLog>(maxBacklog,
+                new GroupingDataflowBlockOptions()
+                {
+                    BoundedCapacity = maxBacklog,
+                    Greedy = true
+                });
+
+            TransformBlock<IEnumerable<FunctionResultLog>, IEnumerable<FunctionResultAggregate>> aggregator =
+                new TransformBlock<IEnumerable<FunctionResultLog>, IEnumerable<FunctionResultAggregate>>(transform: (e) => Aggregate(e));
+
+            ActionBlock<IEnumerable<FunctionResultAggregate>> publisher = new ActionBlock<IEnumerable<FunctionResultAggregate>>(
+                (e) => Publish(e),
+                new ExecutionDataflowBlockOptions()
+                {
+                    MaxDegreeOfParallelism = 1,
+                    BoundedCapacity = 32
+                });
+
+            _disposables = new IDisposable[]
+            {
+                _buffer.LinkTo(_batcher),
+                _batcher.LinkTo(aggregator),
+                aggregator.LinkTo(publisher)
+            };
+
+            _windowTimer = new Timer(async (o) => await FlushAsync(), null, maxFlushInterval, maxFlushInterval);
+        }
+
+        public Task FlushAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            _batcher?.TriggerBatch();
```

Comment:
Eventually, yes, it triggers a publish. More accurately, a `BatchBlock` collects items until it hits the max, at which point it passes a collection of those items onto the next block in the chain. In our case, that's the `TransformBlock`, which does the aggregation. When that's done, it passes its output to the next block, which for us is `ActionBlock` that does the publishing. Calling [`TriggerBatch()`](https://msdn.microsoft.com/en-us/library/hh160310(v=vs.110).aspx) short circuits the "wait for max items" behavior and auto-flushes the items in the `BatchBlock` to the next block, which then eventually get published.

Another pre-emptive question I'll try to answer -- is the `BufferBlock` at the front of the pipeline necessary? It's there in the chance that the Batch block is processing items and not accepting more -- it gives them a place to be stored until the Batch block is ready for them. I wasn't able to get it to break with this configuration so I didn't want to break it down any further.

Take a read through this document -- it gives a pretty good overview: https://msdn.microsoft.com/en-us/library/hh228603(v=vs.110).aspx

@tohling and I were actually discussing some of this the other day -- the Dataflow library is really powerful and provides great patterns like this. But it's really hard to validate the logic just by looking at the code. Even writing tests is unit tests is tough -- I focused on the ends of the pipeline and assumed the other blocks were behaving as they were configured.